client/app/src/main/java/com/microsoft/research/karya/ui/scenarios/common/BaseMTRendererViewModel.kt

```diff
@@ -0,0 +1,324 @@
+
+        if (!File(microtaskTarBallPath).exists()) {
+          inputFileDoesNotExist = true
+          // TODO: Create a MutableLiveData to inform the UI about an alertbox
+        } else {
+          FileUtils.extractGZippedTarBallIntoDirectory(
+            microtaskTarBallPath,
+            microtaskInputDirectory
+          )
+        }
+      }
+
+      if (inputFileDoesNotExist) return@launch
+
+      if (!currentAssignment.output.isJsonNull) {
+        outputData =
+          if (currentAssignment.output.asJsonObject.has("data")) {
+            currentAssignment.output.asJsonObject.getAsJsonObject("data")
+          } else {
+            JsonObject()
+          }
+
+        logs =
+          if (currentAssignment.output.asJsonObject.has("logs")) {
+            currentAssignment.output.asJsonObject.getAsJsonArray("logs")
+          } else {
+            JsonArray()
+          }
+
+        outputFiles =
+          if (currentAssignment.output.asJsonObject.has("files")) {
+            currentAssignment.output.asJsonObject.getAsJsonArray("files")
+          } else {
+            JsonArray()
+          }
+      }
+
+      setupMicrotask()
+    }
+  }
+
+
+  protected fun getRelativePath(s: String): String {
+    return "$fileDirPath/$s"
+  }
+
+  /**
+   * Get the unique file name of the output for current assignment. [params] is a pair of strings: a
+   * file identifier and extension. The file name is usually the current assignmentID appended with
+   * the identifier. The full file name is unique for a unique [params] pair.
+   */
+  private fun getAssignmentFileName(params: Pair<String, String>): String {
+    val identifier = params.first
+    val extension = params.second
+    val assignmentId = microtaskAssignmentIDs[currentAssignmentIndex]
+
+    return if (identifier == "") "$assignmentId.$extension" else "$assignmentId-$identifier.$extension"
+  }
+
+  /** Get the file path for a scratch file for the current assignment and [params] pair */
+  protected fun getAssignmentScratchFilePath(params: Pair<String, String>): String {
+    val dir_path = "$fileDirPath/microtask-assignment-scratch"
+    val dir = File(dir_path)
+    dir.mkdirs()
+    val fileName = getAssignmentFileName(params)
+    return "$dir_path/$fileName"
+  }
+
+  /** Get the file path for an output file for the current assignment and [params] pair */
+  // TODO: Move Scratch File functions to Karya Containers
+  protected fun getAssignmentScratchFile(params: Pair<String, String>): File {
+    val filePath = getAssignmentScratchFilePath(params)
+    val file = File(filePath)
+    if (file.exists()) file.delete()
+    file.createNewFile()
+    return file
+  }
+
+  /** Reset existing microtask. Useful on activity restart. */
+  protected fun resetMicrotask() {
+    getAndSetupMicrotask()
+  }
+
+  @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+  /** Cleanup function called during [onStop]. */
+  protected abstract fun cleanupOnStop() // Set on Base Viewmodel class
```

Comment:
This is wrong and will probably cause issues. `ViewModel` isn't scoped to the `Fragment's` lifecycle. It can and usually does outlive the `Fragment`. You should manually call `cleanup` from `Fragment's` onStop.